model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/organization/InfinispanOrganizationProvider.java

```diff
@@ -0,0 +1,222 @@
+            if (model == null) return null;
+            if (realmCache.getInvalidations().contains(model.getId())) return model;
+            cached = new CachedOrganization(loaded, getRealm(), model);
+            realmCache.getCache().addRevisioned(cached, realmCache.getStartupRevision());
+
+        // no need to check for realm invalidation as IdP changes are handled by events within InfinispanOrganizationProviderFactory
+        } else if (realmCache.getInvalidations().contains(cached.getId())) {
+            return getCacheDelegate(jpaOrgDelegate.getByDomainName(domainName));
+        } else if (managedOrganizations.containsKey(cached.getId())) {
+            return managedOrganizations.get(cached.getId());
+        }
+        OrganizationAdapter adapter = new OrganizationAdapter(cached, realmCache, jpaOrgDelegate);
+        managedOrganizations.put(cacheKey, adapter);
+        return adapter;
+    }
+
+    @Override
+    public Stream<OrganizationModel> getAllStream(String search, Boolean exact, Integer first, Integer max) {
+        // Return cache delegates to ensure cache invalidation during write operations
+        return getCacheDelegates(jpaOrgDelegate.getAllStream(search, exact, first, max));
+    }
+
+    @Override
+    public Stream<OrganizationModel> getAllStream(Map<String, String> attributes, Integer first, Integer max) {
+        // Return cache delegates to ensure cache invalidation during write operations
+        return getCacheDelegates(jpaOrgDelegate.getAllStream(attributes, first, max));
+    }
+
+    @Override
+    public boolean remove(OrganizationModel organization) {
+        registerOrganizationInvalidation(organization.getId());
+        return jpaOrgDelegate.remove(organization);
+    }
+
+    @Override
+    public void removeAll() {
+        //TODO: won't scale, requires a better mechanism for bulk deleting organizations within a realm
+        //this way, all organizations in the realm will be invalidated ... or should it be invalidated whole realm instead?
+        getAllStream().forEach(this::remove);
+    }
+
+    @Override
+    public boolean addMember(OrganizationModel organization, UserModel user) {
+        return jpaOrgDelegate.addMember(organization, user);
+    }
+
+    @Override
+    public boolean removeMember(OrganizationModel organization, UserModel member) {
+        return jpaOrgDelegate.removeMember(organization, member);
+    }
+
+    @Override
+    public Stream<UserModel> getMembersStream(OrganizationModel organization, String search, Boolean exact, Integer first, Integer max) {
+        return jpaOrgDelegate.getMembersStream(organization, search, exact, first, max);
+    }
+
+    @Override
+    public UserModel getMemberById(OrganizationModel organization, String id) {
+        return jpaOrgDelegate.getMemberById(organization, id);
+    }
+
+    @Override
+    public OrganizationModel getByMember(UserModel member) {
```

Comment:
This method is used in called from critical places like during authentication and token mapping. Can we have a cache here too?

Reply:
+1

Reply:
The `id` of the member's organization is stored as attribute of the passed `UserModel`. So it doesn't make much sense to create new cache-able object to cahe this information separately. I can see two ways how to address it.

1. implementation in `InfinispanOrganizationProvider` would look like
```java
    @Override
    public OrganizationModel getByMember(UserModel member) {
        if (member == null) {
            throw new ModelException("Member cannot be null.");
        }

        String orgId = member.getFirstAttribute(ORGANIZATION_ATTRIBUTE);
        if (orgId == null) {
            return null;
        }

        return getById(orgId);
    }
```

2. implementation in `InfinispanOrganizationProvider` would look like
```java
    @Override
    public OrganizationModel getByMember(UserModel member) {
        return orgDelegate.getByMember(member);
    }
```
and `JpaOrganizationProvider` would be changed to
```java
    @Override
    public OrganizationModel getByMember(UserModel member) {
        throwExceptionIfObjectIsNull(member, "User");

        String orgId = member.getFirstAttribute(ORGANIZATION_ATTRIBUTE);

        if (orgId == null) {
            return null;
        }

        // need to go via the session to avoid bypassing the cache
        return session.getProvider(OrganizationProvider.class).getById(orgId);
    }
```

Please let me know what option would be preferred. I tend to incline towards second one as it avoids duplicating code, on the other hand Infinispan provider relies on underlying delegate to avoid bypassing the cache. Or maybe there is another, better way :)

Reply:
True, there is nothing really to cache there. Also agree on the second option.

Reply:
I would also vote for option #2 as I remember a few places where we do that - JPA uses the session to obtain the main object to allow for caching (Groups does it like this, I think)